cloud-init/qemu_performance.py

```diff
@@ -0,0 +1,549 @@
+    )
+    if analyze_times:
+        boot_sample["time_systemd_kernel"] = float(analyze_times["kernel"])
+        boot_sample["time_systemd_userspace"] = float(analyze_times["user"])
+    systemd_blames = {}
+    for line in boot_sample["systemd_analyze_blame"].splitlines():
+        blame = re.match(r"\s*(?P<cost>[\d.]+)s (?P<service_name>\S+)", line)
+        if blame:
+            systemd_blames[blame["service_name"]] = float(blame["cost"])
+    boot_sample["time_systemd_blames"] = systemd_blames
+
+    # Process cloud-init analyze blame output for each config module cost
+    cloudinit_blames = {}
+    for line in boot_sample["cloudinit_analyze_blame"].splitlines():
+        blame = re.match(r"(?P<cost>[^s]+)s \((?P<module_name>[^\)]+)\)", line.strip())
+        if blame:
+            cloudinit_blames[blame["module_name"]] = float(blame["cost"])
+
+    # Process cloud-init analyze show output for each boot stage cost
+    for match in re.findall(
+        r"Finished stage: \((?P<stage>[^)]+)\) (?P<cost>[\d.]+) seconds",
+        boot_sample["cloudinit_analyze"],
+    ):
+        cloudinit_blames[f"stage/{match[0]}"] = float(match[1])
+    stage_total_time = re.match(
+         r".*\nTotal Time: (?P<cost>[\d\.]+) seconds\n",
+         boot_sample["cloudinit_analyze"],
+    re.DOTALL)
+    if stage_total_time:
+        boot_sample["time_cloudinit_total"] = float(stage_total_time["cost"])
+    boot_sample["time_cloudinit_blames"] = cloudinit_blames
+
+
+def update_averages(boot_samples: list[dict], data_dir: Path, data_type: str):
+    avg_data = {
+        "client_time_to_ssh": {"samples": []},
+        "client_time_to_cloudinit_done": {"samples": []},
+        "time_systemd_kernel": {"samples": []},
+        "time_systemd_userspace": {"samples": []},
+        "time_cloudinit_total": {"samples": []},
+        "time_systemd_blames": {},
+        "time_cloudinit_blames": {},
+    }
+    for boot_sample in boot_samples:
+        process_bootspeed_data(boot_sample)
+        for k in avg_data:
+            if "blames" in k:
+                for service_name in boot_sample[k]:
+                    if service_name not in avg_data[k]:
+                        avg_data[k][service_name] = {
+                            "samples": [boot_sample[k][service_name]]
+                        }
+                    else:
+                        avg_data[k][service_name]["samples"].append(
+                            boot_sample[k][service_name]
+                        )
+            else:
+                avg_data[k]["samples"].append(boot_sample[k])
+
+    for k in avg_data:
+        if "blames" in k:
+            for service_name in avg_data[k]:
+                avg_data[k][service_name].update(
+                    get_max_min_avg(avg_data[k][service_name]["samples"])
+                )
+        else:
+            avg_data[k].update(get_max_min_avg(avg_data[k]["samples"]))
+    for idx, boot_sample in enumerate(boot_samples):
```

Comment:
Why is a function called `update_averages()` writing files?

It would be much more intuitive for data to be written after it is collected (up in `inspect_boot_time_deltas_for_deb()`)